rfd/0171-database-session-playback.md

```diff
@@ -0,0 +1,376 @@
+emitted on PostgreSQL when a prepared statement is executed.
+
+The session recordings are available through the `tsh play` command, but they
+can only be viewed as JSON:
+
+
+```json
+[
+  {
+      "ei": 0,
+      "event": "db.session.start",
+      "code": "TDB00I",
+      "success": true,
+      ...
+  },
+  {
+      "ei": 1,
+      "event": "db.session.postgres.statements.execute",
+      "code": "TPG02I",
+      "portal_name": "test",
+      ...
+  },
+  {
+      "ei": 2,
+      "event": "db.session.query",
+      "code": "TDB02I",
+      "db_query": "SELECT 1;",
+      "success": true,
+      ...
+  },
+  {
+      "ei": 3,
+      "event": "db.session.end",
+      "code": "TDB01I",
+      ...
+  }]
+```
+
+### Audit events
+
+To have a complete recording with users input and server response, we'll
+include the command result event:
+
+```proto
+// DatabaseFieldType enum for database data fields data types.
+enum DatabaseDataFieldType {
+  UNSPECIFIED = 0;
+  STRING = 1;
+  NUMBER = 2;
+  DATE = 3;
+  BINARY = 4;
+}
+
+// DatabaseDataFieldMetadata contains metadata for database data fields.
+message DatabaseDataFieldMetadata {
+  // Name is the field name.
+  string Name = 1;
+  // Type is the field type. Since the values are stored inside the event as
+  // strings, this type can convert the row values back into their original
+  // type. This information can be used to format the row values.
+  DatabaseDataFieldType Type = 2;
+}
+
+// DatabaseDataRow represents a single data row, containing a list field values.
+message DatabaseDataRow {
+  // Values the row field values.
+  // Value is the row field values. The values aren not expected to be the
+  // direct database type but their string representation, making it possible to
+  // display them without parsing. For example, binary data can be encoded into
+  // base64.
+  repeated string Values = 1;
+}
+
+// DatabaseResultData database results data, containing fields metadata and also
+// the data itself.
+message DatabaseResultData {
+  // FieldMetadata list of fields metadata.
+  repeated DatabaseDataFieldMetadata Fields = 1;
+  // Rows list of the field values.
+  repeated DatabaseDataRow Rows = 2;
+}
+
+// DatabaseSessionCommandResult represents the result of a user command. It is
+// expected that for each user command/query there will be a corresponding
+// result.
+message DatabaseSessionCommandResult {
+  // Metadata is a common event metadata.
+  Metadata Metadata = 1;
+  // User is a common user event metadata.
+  UserMetadata User = 2;
+  // SessionMetadata is a common event session metadata.
+  SessionMetadata Session = 3;
+  // Database contains database related metadata.
+  DatabaseMetadata Database = 4;
+
+  oneof Result {
+    // Status of the query execution. It is used when the command doesn't return
+    // any data.
+    Status status = 5;
+    // Data contains the fields and data of the query execution.
+    DatabaseResultData Data = 6;
+  }
```

Comment:
Could we do both all the time instead of oneof?

Reply:
Do think there will be any benefit of doing this?